tests/providers/amazon/aws/hooks/test_elasticache_replication_group.py

```diff
@@ -0,0 +1,291 @@
+            replication_group_id=self.REPLICATION_GROUP_ID,
+            max_retries=1,
+            initial_sleep_time=1,  # seconds
+        )
+        assert response is True
+
+    def test_delete_replication_group(self):
+        self._create_replication_group()
+
+        self.hook.conn.delete_replication_group.return_value = {
+            "ReplicationGroup": {
+                "ReplicationGroupId": self.REPLICATION_GROUP_ID,
+                "Status": "deleting"
+            }
+        }
+
+        # Wait for availability, can only delete when replication group is available
+        self.hook.conn.describe_replication_groups.return_value = {
+            "ReplicationGroups": [
+                {
+                    "ReplicationGroupId": self.REPLICATION_GROUP_ID,
+                    "Status": "available"
+                }
+            ]
+        }
+
+        response = self.hook.wait_for_availability(
+            replication_group_id=self.REPLICATION_GROUP_ID,
+            max_retries=1,
+            initial_sleep_time=1,  # seconds
+        )
+        assert response is True
+
+        response = self.hook.delete_replication_group(replication_group_id=self.REPLICATION_GROUP_ID)
+        assert response["ReplicationGroup"]["ReplicationGroupId"] == self.REPLICATION_GROUP_ID
+        assert response["ReplicationGroup"]["Status"] == "deleting"
+
+    # noinspection PyUnusedLocal
+    def _mock_describe_side_effect(self, *args, **kwargs):
+        """
+        Mock describe calls to replication group for testing delete calls
+        """
+        # On first call replication group is in available state, this will allow to initiate a delete
+        # A replication group can only be deleted when it is in `available` state
+        if self.describe_call_count_for_delete == 0:
+            self.describe_call_count_for_delete = self.describe_call_count_for_delete + 1
+
+            return {
+                "ReplicationGroups": [
+                    {
+                        "ReplicationGroupId": self.REPLICATION_GROUP_ID,
+                        "Status": "available"
+                    }
+                ]
+            }
+
+        # On second call replication group is in deleting state
+        if self.describe_call_count_for_delete == 1:
+            self.describe_call_count_for_delete = self.describe_call_count_for_delete + 1
+
+            return {
+                "ReplicationGroups": [
+                    {
+                        "ReplicationGroupId": self.REPLICATION_GROUP_ID,
+                        "Status": "deleting"
+                    }
+                ]
+            }
+
+        # On further calls we will assume the replication group is deleted
+        class MockReplicationGroupNotFoundFault(BaseException):
+            pass
+
+        self.hook.conn.exceptions.ReplicationGroupNotFoundFault = MockReplicationGroupNotFoundFault
+
+        raise self.hook.conn.exceptions.ReplicationGroupNotFoundFault
```

Comment:
You don't need to save `describe_call_count_for_delete`. `side_effect` will do that for you.

Example from https://docs.python.org/3/library/unittest.mock.html#quick-guide:
```python
>>> mock.side_effect = [5, 4, 3, 2, 1]
>>> mock(), mock(), mock()
(5, 4, 3)
```

Reply:
So you could do sth. like this: 
```python
mock.side_effect = [{
    "ReplicationGroups": [{
        "ReplicationGroupId": self.REPLICATION_GROUP_ID,
        "Status": "available"
    }]
}, {
    "ReplicationGroups": [{
        "ReplicationGroupId": self.REPLICATION_GROUP_ID,
        "Status": "deleting"
    }]
}, self.hook.conn.exceptions.ReplicationGroupNotFoundFault()]
```

Reply:
Thanks @feluelle, this makes it better.
The only issue I faced is that the last item has to be an exception of type BaseException so I have created a separate method for that and returning that method from here.